### DISCUSSION TITLE
RFC Tracking Discussion: Hook Service

### DISCUSSION THREAD

**Discussion by Alfred-Mountfield** [blockprotocol/blockprotocol#475] [RFC]
This is the discussion for tracking implementation of #430. Fixes and changes to the phrasing/correctness of the RFC should be submitted as PRs [to the RFC](https://github.com/blockprotocol/blockprotocol/blob/main/rfcs/text/0430-hook-service.md)

**Comment by nathggns** [comment]
I'm currently working on an implementation of the hook service for blocks and embedding applications to use at `@blockprotocol/hook`, which I'll make a PR for soon. 

There'll also be a helpful React hook which has not yet been discussed, which at the moment looks something like this:

```ts
export declare const useHookRef: <T extends HTMLElement>(service: HookBlockHandler | null, type: string, path: string, fallback: (node: T | null) => void) => (node: T | null) => void;
```

This first PR will implement the proposed changes to `@blockprotocol/core`, and it won't add these changes to the specification on our website nor will it add the hook service to the specification.

**Reply by nathggns** [reply]
The API of this changed somewhat but the PR is here https://github.com/blockprotocol/blockprotocol/pull/479

**Comment by CiaranMn** [comment]
I am going to write the specification changes required, i.e.
1. changes to the `core` specification and any docs updates
2. new `hook` service specification and mention in docs
– in that order.

**Comment by CiaranMn** [comment]
The PR introducing the formal spec is at https://github.com/blockprotocol/blockprotocol/pull/676

We decided not to introduce any changes to the `core` specification in the end. The one change suggested in the PR was that apps should respond with a `NOT_IMPLEMENTED` error if they did not implement a service, but this would have been indistinguishable from the proposed `NOT_IMPLEMENTED` error under the hook service if an application didn't implement a particular hook `type`.

We should think separately about how blocks can statically declare their service requirements, to avoid apps even attempting to load ones which rely on a service which is not implemented. As part of that we can add runtime errors if a service is not implemented.
